File: python/prism_sdk/statistical_analysis_workflow_fabric.py
"""Federated continual statistical/causal/ML workflow fabric for ``AFA-examples-P13-F16``."""
from __future__ import annotations

from dataclasses import asdict, dataclass
import hashlib
import json
import re
from typing import Any, Mapping

from .research_contracts import PRECLINICAL_BOUNDARY, RESEARCH_CONTRACT_SCHEMA_VERSION, ResearchContractError

FEATURE_ID = "AFA-examples-P13-F16"
CONTRACT_VERSION = "examples-federated-continual-statistical-analysis-workflow-fabric/1.0"
INPUT_SCHEMA = "AnalysisWorkflowDraft5@1"
OUTPUT_SCHEMA = "AnalysisWorkflowRun8@1"
STAGES = ("admit", "checkpoint", "validate", "schedule", "retain-receipt")


def _hash(value: Any) -> str:
    return hashlib.sha256(json.dumps(value, sort_keys=True, separators=(",", ":"), ensure_ascii=False).encode()).hexdigest()


def _digest(value: Any) -> bool:
    return isinstance(value, str) and re.fullmatch(r"[0-9a-f]{64}", value) is not None


def _canonical(values: list[str] | tuple[str, ...]) -> bool:
    return list(values) == sorted(set(values))


@dataclass(frozen=True)
class StatisticalAnalysisWorkflowRun:
    schema_version: str; contract_version: str; feature_id: str; request_id: str; federation_id: str; purpose: str; semantic_profile: str; disposition: str; stage_order: tuple[str, ...]; required_candidate_order: tuple[str, ...]; selected_candidate_order: tuple[str, ...]; pending_candidate_order: tuple[str, ...]; blocked_candidate_order: tuple[str, ...]; compensated_candidate_order: tuple[str, ...]; omission_order: tuple[str, ...]; uncertainty_order: tuple[str, ...]; negative_evidence_order: tuple[str, ...]; checkpoint_digest: str; replay_identity: str; workflow_digest: str; artifact: Mapping[str, Any]; effect_receipts: tuple[str, ...]; raw_data_local: bool; aggregate_only: bool; boundary: str

    def validate(self) -> None:
        if (self.schema_version, self.contract_version, self.feature_id) != (RESEARCH_CONTRACT_SCHEMA_VERSION, CONTRACT_VERSION, FEATURE_ID) or self.boundary != PRECLINICAL_BOUNDARY or self.raw_data_local is not True or self.aggregate_only is not True or tuple(self.stage_order) != STAGES or not self.required_candidate_order or not self.effect_receipts or not all(isinstance(value, str) and value.strip() for value in (self.request_id, self.federation_id, self.purpose, self.semantic_profile)):
            raise ResearchContractError("analysis workflow identity, stages, locality, aggregate boundary, or effects are incomplete")
        for values in (self.required_candidate_order, self.selected_candidate_order, self.pending_candidate_order, self.blocked_candidate_order, self.compensated_candidate_order, self.omission_order, self.uncertainty_order, self.negative_evidence_order, self.effect_receipts):
            if not _canonical(values): raise ResearchContractError("analysis workflow ordering is not canonical")
        required = set(self.required_candidate_order); states = list(self.selected_candidate_order) + list(self.pending_candidate_order) + list(self.blocked_candidate_order)
        if len(required) != len(self.required_candidate_order) or len(states) != len(required) or set(states) != required or any(item not in required for item in self.compensated_candidate_order): raise ResearchContractError("analysis candidate states do not partition the required plan")
        for digest in (self.checkpoint_digest, self.replay_identity, self.workflow_digest, self.artifact.get("content_hash")):
            if not _digest(digest): raise ResearchContractError("analysis workflow digest is invalid")
        if self.artifact.get("content_type") != "application/vnd.aurora.analysis-workflow-run+json": raise ResearchContractError("analysis workflow artifact type is invalid")
        if any(not effect.startswith("schedule:analysis-workflow:") and not effect.startswith("compensate:analysis-workflow:") and effect != "block:unsafe-release" for effect in self.effect_receipts): raise ResearchContractError("analysis workflow effect is outside schedule/compensation gate")

    def to_dict(self) -> dict[str, Any]:
        value = asdict(self)
        for key, item in value.items():
            if isinstance(item, tuple): value[key] = list(item)
        return value


def assure_statistical_analysis_workflow(*, draft: Mapping[str, Any]) -> StatisticalAnalysisWorkflowRun:
    required = ("request_id", "federation_id", "purpose", "semantic_profile")
    if any(not str(draft.get(field, "")).strip() for field in required) or not draft.get("required_candidate_order") or not _canonical([str(item) for item in draft["required_candidate_order"]]) or tuple(draft.get("stage_order", ())) != STAGES or draft.get("raw_data_local") is not True or draft.get("aggregate_only") is not True or int(draft.get("budget_units", 0)) <= 0 or int(draft.get("budget_units", 0)) > int(draft.get("max_budget_units", 0)) or draft.get("boundary") != PRECLINICAL_BOUNDARY or not _digest(draft.get("replay_identity")):
        raise ResearchContractError("analysis workflow request identity, stage protocol, locality, aggregate boundary, budget, replay, or boundary is invalid")
    required_order = [str(item) for item in draft["required_candidate_order"]]
    if len(set(required_order)) != len(required_order): raise ResearchContractError("required analysis candidates must be unique")
    candidates = {}; selected: set[str] = set(); pending: set[str] = set(); blocked: set[str] = set(); omissions: set[str] = set(); uncertainty: set[str] = set(); negative: set[str] = set()
    for candidate in draft.get("candidates", []):
        cid = str(candidate.get("candidate_id", "")); omissions_list = [str(item) for item in candidate.get("omissions", [])]; uncertainty_list = [str(item) for item in candidate.get("uncertainty", [])]
        if not cid.strip() or cid not in set(required_order) or cid in candidates or not _canonical(omissions_list) or not _canonical(uncertainty_list) or not str(candidate.get("estimand", "")).strip() or not str(candidate.get("method_family", "")).strip() or candidate.get("semantic_profile") != draft["semantic_profile"] or not _digest(candidate.get("baseline_digest")) or not _digest(candidate.get("provenance_digest")) or candidate.get("local_data") is not True or int(candidate.get("quality_score_milli", 0)) > 1000:
            raise ResearchContractError("analysis candidate identity, evidence, baseline, provenance, profile, quality, or locality is invalid")
        candidates[cid] = candidate; omissions.update(f"candidate:{cid}:{item}" for item in omissions_list); uncertainty.update(f"candidate:{cid}:{item}" for item in uncertainty_list)
        if candidate.get("negative_result"): negative.add(f"candidate:{cid}:negative-result")
        state = str(candidate.get("evidence_state", "unknown"))
        if state in {"proven", "supported"} and int(candidate.get("quality_score_milli", 0)) >= 700 and not omissions_list and not uncertainty_list: selected.add(cid)
        elif state == "contradicted": blocked.add(cid); negative.add(f"candidate:{cid}:contradicted")
        else: pending.add(cid); uncertainty.add(f"candidate:{cid}:not-qualified")
    for cid in set(required_order) - set(candidates): pending.add(cid); omissions.add(f"missing-candidate:{cid}"); uncertainty.add(f"missing-candidate:{cid}")
    violations: set[str] = set()
    for name, field in (("policy", "policy_allow"), ("protected-closure", "protected_closure"), ("signed-approval", "signed_approval"), ("federation-approval", "federation_approved")):
        if draft.get(field) is not True: violations.add(name)
    for event in draft.get("adversarial_events", []): violations.add(f"adversarial:{event}"); omissions.add(f"workflow:adversarial:{event}")
    disposition = "blocked" if violations or draft.get("adversarial_events") else "partial" if pending or blocked or uncertainty else "qualified"
    if disposition == "blocked": blocked = set(required_order); selected.clear(); pending.clear()
    selected_order = sorted(selected); pending_order = sorted(pending); blocked_order = sorted(blocked)
    checkpoint = {"request_id": draft["request_id"], "stage_order": list(STAGES), "required_candidate_order": required_order, "selected_candidate_order": selected_order, "pending_candidate_order": pending_order, "blocked_candidate_order": blocked_order, "replay_identity": draft["replay_identity"]}
    checkpoint_digest = _hash(checkpoint); workflow_digest = _hash({"checkpoint_digest": checkpoint_digest, "semantic_profile": draft["semantic_profile"], "disposition": disposition}); artifact = {"schema_version": RESEARCH_CONTRACT_SCHEMA_VERSION, "artifact_id": f"examples-analysis-workflow:{draft['request_id']}", "content_type": "application/vnd.aurora.analysis-workflow-run+json", "content_hash": checkpoint_digest, "semantic_loss": [], "provenance": [{"source_id": draft["federation_id"], "relation": "analysis-workflow-checkpoint", "digest": checkpoint_digest}], "boundary": PRECLINICAL_BOUNDARY}; receipt = StatisticalAnalysisWorkflowRun(RESEARCH_CONTRACT_SCHEMA_VERSION, CONTRACT_VERSION, FEATURE_ID, str(draft["request_id"]), str(draft["federation_id"]), str(draft["purpose"]), str(draft["semantic_profile"]), disposition, STAGES, tuple(required_order), tuple(selected_order), tuple(pending_order), tuple(blocked_order), tuple(pending_order), tuple(sorted(omissions)), tuple(sorted(uncertainty)), tuple(sorted(negative)), checkpoint_digest, str(draft["replay_identity"]), workflow_digest, artifact, (f"schedule:analysis-workflow:{draft['request_id']}",) if disposition == "qualified" else (f"compensate:analysis-workflow:{draft['request_id']}",) if disposition == "partial" else ("block:unsafe-release",), True, True, PRECLINICAL_BOUNDARY); receipt.validate(); return receipt


__all__ = ["FEATURE_ID", "CONTRACT_VERSION", "INPUT_SCHEMA", "OUTPUT_SCHEMA", "StatisticalAnalysisWorkflowRun", "assure_statistical_analysis_workflow"]

File: crates/examples/src/statistical_analysis_workflow_fabric.rs
//! Federated continual statistical/causal/ML analysis workflow fabric for Examples.
//!
//! Atlas feature: `AFA-examples-P13-F16`.
//!
//! The fabric compiles caller-supplied analysis attestations into a resumable workflow receipt.
//! It does not fit models, execute notebooks, access clinical data, or move raw preclinical
//! observations; a separately governed executor may consume a qualified plan.

use bioprism_foundation::{
    AuthorityRequirement, AutonomyTier, CapabilityManifest, Determinism, Effect, EvidenceReference,
    EvidenceState, ProvenanceLink, ResearchSurface, SemanticLoss, TypedPort, TypedResearchArtifact,
    PRECLINICAL_BOUNDARY, RESEARCH_CONTRACT_SCHEMA_VERSION,
};
use bioprism_ids::ContentHash;
use serde::{Deserialize, Serialize};
use serde_json::json;
use std::collections::BTreeSet;
use thiserror::Error;

pub const FEATURE_ID: &str = "AFA-examples-P13-F16";
pub const CONTRACT_VERSION: &str =
    "examples-federated-continual-statistical-analysis-workflow-fabric/1.0";
pub const INPUT_SCHEMA: &str = "AnalysisWorkflowDraft5@1";
pub const OUTPUT_SCHEMA: &str = "AnalysisWorkflowRun8@1";

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct AnalysisCandidate {
    pub candidate_id: String,
    pub estimand: String,
    pub method_family: String,
    pub input_digest: ContentHash,
    pub output_digest: ContentHash,
    pub baseline_digest: Option<ContentHash>,
    pub provenance_digest: Option<ContentHash>,
    pub replay_identity: ContentHash,
    pub semantic_profile: String,
    pub evidence_state: EvidenceState,
    pub quality_score_milli: u32,
    pub local_data: bool,
    pub omissions: Vec<String>,
    pub uncertainty: Vec<String>,
    pub negative_result: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct AnalysisWorkflowDraft {
    pub request_id: String,
    pub federation_id: String,
    pub purpose: String,
    pub semantic_profile: String,
    pub required_candidate_order: Vec<String>,
    pub candidates: Vec<AnalysisCandidate>,
    pub stage_order: Vec<String>,
    pub replay_identity: ContentHash,
    pub policy_allow: bool,
    pub protected_closure: bool,
    pub signed_approval: bool,
    pub federation_approved: bool,
    pub raw_data_local: bool,
    pub aggregate_only: bool,
    pub budget_units: u32,
    pub max_budget_units: u32,
    pub adversarial_events: Vec<String>,
    pub boundary: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct AnalysisWorkflowRun {
    pub schema_version: String,
    pub contract_version: String,
    pub feature_id: String,
    pub request_id: String,
    pub federation_id: String,
    pub purpose: String,
    pub semantic_profile: String,
    pub disposition: String,
    pub stage_order: Vec<String>,
    pub required_candidate_order: Vec<String>,
    pub selected_candidate_order: Vec<String>,
    pub pending_candidate_order: Vec<String>,
    pub blocked_candidate_order: Vec<String>,
    pub compensated_candidate_order: Vec<String>,
    pub omission_order: Vec<String>,
    pub uncertainty_order: Vec<String>,
    pub negative_evidence_order: Vec<String>,
    pub checkpoint_digest: ContentHash,
    pub replay_identity: ContentHash,
    pub workflow_digest: ContentHash,
    pub artifact: TypedResearchArtifact,
    pub effect_receipts: Vec<String>,
    pub raw_data_local: bool,
    pub aggregate_only: bool,
    pub boundary: String,
}

#[derive(Debug, Error, PartialEq, Eq)]
pub enum AnalysisWorkflowError {
    #[error("invalid statistical analysis workflow draft: {0}")]
    Invalid(String),
    #[error("statistical analysis workflow artifact failed: {0}")]
    Artifact(String),
}

fn canonical(values: &[String]) -> bool {
    values.windows(2).all(|pair| pair[0] < pair[1])
}

impl AnalysisWorkflowRun {
    pub fn validate(&self) -> Result<(), AnalysisWorkflowError> {
        if self.schema_version != RESEARCH_CONTRACT_SCHEMA_VERSION
            || self.contract_version != CONTRACT_VERSION
            || self.feature_id != FEATURE_ID
            || self.boundary != PRECLINICAL_BOUNDARY
            || !self.raw_data_local
            || !self.aggregate_only
            || self.request_id.trim().is_empty()
            || self.federation_id.trim().is_empty()
            || self.purpose.trim().is_empty()
            || self.semantic_profile.trim().is_empty()
            || self.stage_order.is_empty()
            || self.required_candidate_order.is_empty()
            || self.effect_receipts.is_empty()
        {
            return Err(AnalysisWorkflowError::Invalid(
                "analysis workflow identity, stages, candidates, locality, aggregate boundary, or effects are incomplete".into(),
            ));
        }
        for values in [
            &self.required_candidate_order,
            &self.selected_candidate_order,
            &self.pending_candidate_order,
            &self.blocked_candidate_order,
            &self.compensated_candidate_order,
            &self.omission_order,
            &self.uncertainty_order,
            &self.negative_evidence_order,
            &self.effect_receipts,
        ] {
            if !canonical(values) {
                return Err(AnalysisWorkflowError::Invalid(
                    "analysis workflow ordering is not canonical".into(),
                ));
            }
        }
        if self.stage_order
            != vec![
                "admit".to_string(),
                "checkpoint".to_string(),
                "validate".to_string(),
                "schedule".to_string(),
                "retain-receipt".to_string(),
            ]
        {
            return Err(AnalysisWorkflowError::Invalid(
                "analysis workflow stage protocol is invalid".into(),
            ));
        }
        let required = self
            .required_candidate_order
            .iter()
            .cloned()
            .collect::<BTreeSet<_>>();
        let states = self
            .selected_candidate_order
            .iter()
            .chain(self.pending_candidate_order.iter())
            .chain(self.blocked_candidate_order.iter())
            .cloned()
            .collect::<Vec<_>>();
        if required.len() != self.required_candidate_order.len()
            || states.len() != required.len()
            || states.iter().cloned().collect::<BTreeSet<_>>() != required
            || self
                .compensated_candidate_order
                .iter()
                .any(|candidate| !required.contains(candidate))
        {
            return Err(AnalysisWorkflowError::Invalid(
                "analysis candidate states do not partition the required plan".into(),
            ));
        }
        if self.effect_receipts.iter().any(|effect| {
            !effect.starts_with("schedule:analysis-workflow:")
                && !effect.starts_with("compensate:analysis-workflow:")
                && effect != "block:unsafe-release"
        }) {
            return Err(AnalysisWorkflowError::Invalid(
                "analysis workflow effect is outside schedule/compensation gate".into(),
            ));
        }
        self.artifact
            .validate_metadata()
            .map_err(|error| AnalysisWorkflowError::Artifact(error.to_string()))
    }

    pub fn digest(&self) -> Result<ContentHash, AnalysisWorkflowError> {
        self.validate()?;
        ContentHash::of_value(
            &serde_json::to_value(self)
                .map_err(|error| AnalysisWorkflowError::Artifact(error.to_string()))?,
        )
        .map_err(|error| AnalysisWorkflowError::Artifact(error.to_string()))
    }
}

pub fn capability_manifest() -> CapabilityManifest {
    CapabilityManifest {
        schema_version: RESEARCH_CONTRACT_SCHEMA_VERSION.into(),
        capability_id: FEATURE_ID.into(),
        version: CONTRACT_VERSION.into(),
        owner_crate: "examples".into(),
        consumers: BTreeSet::from([
            "research program lead".into(),
            "analysis workflow operator".into(),
            "federated evaluation gateway".into(),
        ]),
        behavior: "compiles typed statistical, causal, and ML analysis attestations into a resumable checkpointed workflow plan without executing models".into(),
        value: "makes analysis selection, evidence closure, compensation, and federation readiness auditable before computation".into(),
        inputs: vec![TypedPort {
            name: "analysis_workflow_draft".into(),
            schema: INPUT_SCHEMA.into(),
            required: true,
        }],
        outputs: vec![TypedPort {
            name: "analysis_workflow_run".into(),
            schema: OUTPUT_SCHEMA.into(),
            required: true,
        }],
        effects: BTreeSet::from([
            Effect::ReadLocalData,
            Effect::WriteLocalArtifact,
            Effect::FederationExport,
        ]),
        permissions: BTreeSet::from([
            "schedule:analysis-workflows".into(),
            "exchange:aggregate-analysis-manifests".into(),
        ]),
        determinism: Determinism::ByteStable,
        evidence: vec![
            EvidenceReference {
                source_id: "w3c-prov-o".into(),
                state: EvidenceState::Supported,
                locator: Some("https://www.w3.org/TR/prov-o/".into()),
            },
            EvidenceReference {
                source_id: "cwl".into(),
                state: EvidenceState::Supported,
                locator: Some("https://www.commonwl.org/specification/".into()),
            },
            EvidenceReference {
                source_id: "opentelemetry".into(),
                state: EvidenceState::Supported,
                locator: Some("https://opentelemetry.io/docs/specs/".into()),
            },
        ],
        authority_requirements: vec![AuthorityRequirement {
            role: "analysis workflow operator".into(),
            reason: "approve scheduling and aggregate-only federation of analysis manifests".into(),
        }],
        autonomy_tier: AutonomyTier::A2,
        surfaces: BTreeSet::from([
            ResearchSurface::Api,
            ResearchSurface::Sdk,
            ResearchSurface::Protocol,
            ResearchSurface::Policy,
            ResearchSurface::Operator,
        ]),
        boundary: PRECLINICAL_BOUNDARY.into(),
    }
}

pub fn assure(draft: &AnalysisWorkflowDraft) -> Result<AnalysisWorkflowRun, AnalysisWorkflowError> {
    if draft.request_id.trim().is_empty()
        || draft.federation_id.trim().is_empty()
        || draft.purpose.trim().is_empty()
        || draft.semantic_profile.trim().is_empty()
        || draft.required_candidate_order.is_empty()
        || !canonical(&draft.required_candidate_order)
        || draft.stage_order
            != vec![
                "admit".to_string(),
                "checkpoint".to_string(),
                "validate".to_string(),
                "schedule".to_string(),
                "retain-receipt".to_string(),
            ]
        || !draft.raw_data_local
        || !draft.aggregate_only
        || draft.budget_units == 0
        || draft.max_budget_units == 0
        || draft.budget_units > draft.max_budget_units
        || draft.boundary != PRECLINICAL_BOUNDARY
    {
        return Err(AnalysisWorkflowError::Invalid(
            "analysis workflow identity, stages, locality, aggregate boundary, or budget is invalid".into(),
        ));
    }
    let required = draft
        .required_candidate_order
        .iter()
        .cloned()
        .collect::<BTreeSet<_>>();
    if required.len() != draft.required_candidate_order.len() {
        return Err(AnalysisWorkflowError::Invalid(
            "required analysis candidates must be unique".into(),
        ));
    }
    let mut candidates = BTreeSet::new();
    let mut selected = BTreeSet::new();
    let mut pending = BTreeSet::new();
    let mut blocked = BTreeSet::new();
    let mut omissions = BTreeSet::new();
    let mut uncertainty = BTreeSet::new();
    let mut negative = BTreeSet::new();
    for candidate in &draft.candidates {
        if candidate.candidate_id.trim().is_empty()
            || !required.contains(&candidate.candidate_id)
            || !canonical(&candidate.omissions)
            || !canonical(&candidate.uncertainty)
            || candidate.estimand.trim().is_empty()
            || candidate.method_family.trim().is_empty()
            || candidate.semantic_profile != draft.semantic_profile
            || candidate.baseline_digest.is_none()
            || candidate.provenance_digest.is_none()
            || !candidate.local_data
            || candidate.quality_score_milli > 1000
            || !candidates.insert(candidate.candidate_id.clone())
        {
            return Err(AnalysisWorkflowError::Invalid(
                "analysis candidate identity, evidence, baseline, provenance, profile, quality, or locality is invalid".into(),
            ));
        }
        for item in &candidate.omissions {
            omissions.insert(format!("candidate:{}:{item}", candidate.candidate_id));
        }
        for item in &candidate.uncertainty {
            uncertainty.insert(format!("candidate:{}:{item}", candidate.candidate_id));
        }
        if candidate.negative_result {
            negative.insert(format!(
                "candidate:{}:negative-result",
                candidate.candidate_id
            ));
        }
        match candidate.evidence_state {
            EvidenceState::Proven | EvidenceState::Supported
                if candidate.quality_score_milli >= 700
                    && candidate.omissions.is_empty()
                    && candidate.uncertainty.is_empty() =>
            {
                selected.insert(candidate.candidate_id.clone());
            }
            EvidenceState::Contradicted => {
                blocked.insert(candidate.candidate_id.clone());
                negative.insert(format!("candidate:{}:contradicted", candidate.candidate_id));
            }
            _ => {
                pending.insert(candidate.candidate_id.clone());
                uncertainty.insert(format!(
                    "candidate:{}:not-qualified",
                    candidate.candidate_id
                ));
            }
        }
    }
    for candidate in required.difference(&candidates) {
        pending.insert((*candidate).clone());
        omissions.insert(format!("missing-candidate:{candidate}"));
        uncertainty.insert(format!("missing-candidate:{candidate}"));
    }
    if selected.len() + pending.len() + blocked.len() != required.len() {
        return Err(AnalysisWorkflowError::Invalid(
            "analysis candidate outcomes do not partition required candidates".into(),
        ));
    }
    let mut violations = BTreeSet::new();
    if !draft.policy_allow {
        violations.insert("policy".into());
    }
    if !draft.protected_closure {
        violations.insert("protected-closure".into());
    }
    if !draft.signed_approval {
        violations.insert("signed-approval".into());
    }
    if !draft.federation_approved {
        violations.insert("federation-approval".into());
    }
    for event in &draft.adversarial_events {
        violations.insert(format!("adversarial:{event}"));
        omissions.insert(format!("workflow:adversarial:{event}"));
    }
    let global_block = !violations.is_empty() || !draft.adversarial_events.is_empty();
    let disposition = if global_block {
        "blocked"
    } else if !pending.is_empty() || !blocked.is_empty() || !uncertainty.is_empty() {
        "partial"
    } else {
        "qualified"
    };
    if global_block {
        blocked.extend(required.iter().cloned());
        selected.clear();
        pending.clear();
    }
    let selected_order = selected.into_iter().collect::<Vec<_>>();
    let pending_order = pending.into_iter().collect::<Vec<_>>();
    let blocked_order = blocked.into_iter().collect::<Vec<_>>();
    let compensated_order = pending_order.clone();
    let checkpoint_payload = json!({
        "request_id": draft.request_id,
        "stage_order": draft.stage_order,
        "required_candidate_order": draft.required_candidate_order,
        "selected_candidate_order": selected_order,
        "pending_candidate_order": pending_order,
        "blocked_candidate_order": blocked_order,
        "replay_identity": draft.replay_identity,
    });
    let checkpoint_digest = ContentHash::of_value(&checkpoint_payload)
        .map_err(|error| AnalysisWorkflowError::Artifact(error.to_string()))?;
    let workflow_digest = ContentHash::of_value(&json!({
        "checkpoint_digest": checkpoint_digest,
        "semantic_profile": draft.semantic_profile,
        "disposition": disposition,
    }))
    .map_err(|error| AnalysisWorkflowError::Artifact(error.to_string()))?;
    let artifact = TypedResearchArtifact::from_payload(
        format!("examples-analysis-workflow:{}", draft.request_id),
        "application/vnd.aurora.analysis-workflow-run+json",
        &checkpoint_payload,
        Vec::<SemanticLoss>::new(),
        vec![ProvenanceLink {
            source_id: draft.federation_id.clone(),
            relation: "analysis-workflow-checkpoint".into(),
            digest: checkpoint_digest.clone(),
        }],
    )
    .map_err(|error| AnalysisWorkflowError::Artifact(error.to_string()))?;
    let effect_receipts = if disposition == "qualified" {
        vec![format!("schedule:analysis-workflow:{}", draft.request_id)]
    } else if disposition == "partial" {
        vec![format!("compensate:analysis-workflow:{}", draft.request_id)]
    } else {
        vec!["block:unsafe-release".into()]
    };
    let run = AnalysisWorkflowRun {
        schema_version: RESEARCH_CONTRACT_SCHEMA_VERSION.into(),
        contract_version: CONTRACT_VERSION.into(),
        feature_id: FEATURE_ID.into(),
        request_id: draft.request_id.clone(),
        federation_id: draft.federation_id.clone(),
        purpose: draft.purpose.clone(),
        semantic_profile: draft.semantic_profile.clone(),
        disposition: disposition.into(),
        stage_order: draft.stage_order.clone(),
        required_candidate_order: draft.required_candidate_order.clone(),
        selected_candidate_order: selected_order,
        pending_candidate_order: pending_order,
        blocked_candidate_order: blocked_order,
        compensated_candidate_order: compensated_order,
        omission_order: omissions.into_iter().collect(),
        uncertainty_order: uncertainty.into_iter().collect(),
        negative_evidence_order: negative.into_iter().collect(),
        checkpoint_digest,
        replay_identity: draft.replay_identity.clone(),
        workflow_digest,
        artifact,
        effect_receipts,
        raw_data_local: draft.raw_data_local,
        aggregate_only: draft.aggregate_only,
        boundary: PRECLINICAL_BOUNDARY.into(),
    };
    run.validate()?;
    Ok(run)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn hash() -> ContentHash {
        ContentHash::of_bytes(b"examples-analysis-workflow")
    }

    fn candidate(id: &str, state: EvidenceState) -> AnalysisCandidate {
        AnalysisCandidate {
            candidate_id: id.into(),
            estimand: "effect-on-organoid-growth".into(),
            method_family: "doubly-robust".into(),
            input_digest: hash(),
            output_digest: hash(),
            baseline_digest: Some(hash()),
            provenance_digest: Some(hash()),
            replay_identity: hash(),
            semantic_profile: "analysis:v1".into(),
            evidence_state: state,
            quality_score_milli: 900,
            local_data: true,
            omissions: Vec::new(),
            uncertainty: Vec::new(),
            negative_result: false,
        }
    }

    fn draft() -> AnalysisWorkflowDraft {
        AnalysisWorkflowDraft {
            request_id: "request:examples-analysis".into(),
            federation_id: "federation:analysis".into(),
            purpose: "preclinical-replication".into(),
            semantic_profile: "analysis:v1".into(),
            required_candidate_order: vec!["analysis-a".into(), "analysis-b".into()],
            candidates: vec![
                candidate("analysis-a", EvidenceState::Supported),
                candidate("analysis-b", EvidenceState::Proven),
            ],
            stage_order: vec![
                "admit".into(),
                "checkpoint".into(),
                "validate".into(),
                "schedule".into(),
                "retain-receipt".into(),
            ],
            replay_identity: hash(),
            policy_allow: true,
            protected_closure: true,
            signed_approval: true,
            federation_approved: true,
            raw_data_local: true,
            aggregate_only: true,
            budget_units: 10,
            max_budget_units: 10,
            adversarial_events: Vec::new(),
            boundary: PRECLINICAL_BOUNDARY.into(),
        }
    }

    #[test]
    fn qualified_workflow_schedules() {
        let run = assure(&draft()).unwrap();
        assert_eq!(run.disposition, "qualified");
        assert_eq!(
            run.selected_candidate_order,
            vec!["analysis-a", "analysis-b"]
        );
        assert!(run.effect_receipts[0].starts_with("schedule:analysis-workflow:"));
    }
    #[test]
    fn unknown_candidate_is_partial_and_compensated() {
        let mut value = draft();
        value.candidates[0].evidence_state = EvidenceState::Unknown;
        let run = assure(&value).unwrap();
        assert_eq!(run.disposition, "partial");
        assert!(run.pending_candidate_order.contains(&"analysis-a".into()));
        assert!(run.effect_receipts[0].starts_with("compensate:analysis-workflow:"));
    }
    #[test]
    fn missing_candidate_is_partial() {
        let mut value = draft();
        value.candidates.pop();
        let run = assure(&value).unwrap();
        assert_eq!(run.disposition, "partial");
        assert!(run
            .omission_order
            .iter()
            .any(|item| item.contains("missing-candidate")));
    }
    #[test]
    fn contradiction_blocks() {
        let mut value = draft();
        value.candidates[0].evidence_state = EvidenceState::Contradicted;
        let run = assure(&value).unwrap();
        assert_eq!(run.disposition, "partial");
        assert!(run.blocked_candidate_order.contains(&"analysis-a".into()));
    }
    #[test]
    fn policy_and_adversarial_events_block() {
        let mut value = draft();
        value.policy_allow = false;
        value.adversarial_events = vec!["poisoned-model-card".into()];
        let run = assure(&value).unwrap();
        assert_eq!(run.disposition, "blocked");
        assert_eq!(run.effect_receipts, vec!["block:unsafe-release"]);
    }
    #[test]
    fn manifest_is_a2_and_federated() {
        let manifest = capability_manifest();
        assert_eq!(manifest.autonomy_tier, AutonomyTier::A2);
        assert!(manifest.effects.contains(&Effect::FederationExport));
    }
}
